extensions/kafka-streams/deployment/src/main/java/io/quarkus/kafka/streams/deployment/KafkaStreamsProcessor.java

```diff
@@ -66,20 +62,89 @@ void build(RecorderContext recorder,
+    }
+
+    private void registerClassesThatClientMaySpecify(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        registerExceptionHandler(reflectiveClasses);
+        registerDefaultSerdes(reflectiveClasses);
+    }
+
+    private void registerExceptionHandler(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        Properties properties = buildKafkaStreamsProperties();
+        String exceptionHandlerClassName = properties
+                .getProperty(StreamsConfig.DEFAULT_DESERIALIZATION_EXCEPTION_HANDLER_CLASS_CONFIG);
+
+        if (exceptionHandlerClassName == null)
+            registerDefaultExceptionHandler(reflectiveClasses);
+        else
+            registerClassName(reflectiveClasses, exceptionHandlerClassName);
+    }
+
+    private void registerDefaultExceptionHandler(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        reflectiveClasses.produce(new ReflectiveClassBuildItem(true, false, false, LogAndFailExceptionHandler.class));
+    }
+
+    private void registerDefaultSerdes(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        Properties properties = buildKafkaStreamsProperties();
+        String defaultKeySerdeClass = properties.getProperty(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG);
+        String defaultValueSerdeClass = properties.getProperty(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG);
+
+        if (defaultKeySerdeClass != null)
+            registerClassName(reflectiveClasses, defaultKeySerdeClass);
+        if (defaultValueSerdeClass != null)
+            registerClassName(reflectiveClasses, defaultValueSerdeClass);
+        if (allDefaultSerdesAreNotDefinedInProperties(defaultKeySerdeClass, defaultValueSerdeClass)) {
+            registerDefaultSerde(reflectiveClasses);
+        }
+    }
+
+    private void registerClassName(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses, String defaultKeySerdeClass) {
+        reflectiveClasses.produce(new ReflectiveClassBuildItem(true, false, false, defaultKeySerdeClass));
+    }
+
+    private boolean allDefaultSerdesAreNotDefinedInProperties(String defaultKeySerdeClass, String defaultValueSerdeClass) {
+        return defaultKeySerdeClass == null || defaultValueSerdeClass == null;
+    }
+
+    private void registerDefaultSerde(BuildProducer<ReflectiveClassBuildItem> reflectiveClasses) {
+        reflectiveClasses.produce(new ReflectiveClassBuildItem(true, false, false, ByteArraySerde.class));
+    }
+
     @BuildStep
     @Record(ExecutionTime.STATIC_INIT)
     BeanContainerListenerBuildItem processBuildTimeConfig(KafkaStreamsRecorder recorder) {
-        Config config = ConfigProvider.getConfig();
+        Properties kafkaStreamsProperties = buildKafkaStreamsProperties();
+        return new BeanContainerListenerBuildItem(recorder.configure(kafkaStreamsProperties));
+    }
 
-        Properties properties = new Properties();
+    private Properties buildKafkaStreamsProperties() {
```

Comment:
This one is called three times overall during the build phase. This might be avoided?

Reply:
Does this have a performance impact? 
I imagine the alternative would be to cache the properties in a field. Is this approach preferred?

Reply:
I'd suggest to just build the properties once and pass them down as a parameter to all the methods needing it. This will require to merge the two static init methods that currently require it into one.

Reply:
Ok.... for that, I need to understand where the `KafkaStreamsRecorder` comes from which is passed in to `processBuildTimeConfig`. I tried adding it in as an argument to the `build` method, and it worked. I guess there is some DI involved somewhere?

The other side effect is, obviously, that we're adding another argument, which IMHO is never good.   Would there be any harm, in your opinion, to using fields? The `reflectiveClasses` variable could also benefit from being a field. I'm not familiar with how the class is used.

Reply:
> I tried adding it in as an argument to the build method, and it worked. I guess there is some DI involved somewhere?

Yes, Quarkus will pass all the required parameters for the build step methods automatically.

> Would there be any harm, in your opinion, to using fields?

I don't know tbh., I'm not sure about the lifecycle of these classes (perhaps @gsmet can comment); in any case just a parameter seemed saver and the easiest to me.

Reply:
yes, use a parameter.

Reply:
Thanks. OK - so unless I hear that fields are safe to use, I'll use arguments.
Merging the 2 static init methods nicely is now leading me to extract some methods to indicate whats going on. Can you tell me what these 2 lines from the `build` method do?:
`feature.produce(new FeatureBuildItem(FeatureBuildItem.KAFKA_STREAMS));`
`jni.produce(new JniBuildItem());`

Reply:
This enable JNI for the native build. This is needed for RocksDB which loads native libraries via JNI.

Reply:
Thanks!

Reply:
@gunnarmorling, @gsmet  So, the build failed. After digging about a bit it appears that (at least for the tests) I can't add the `KafkaStreamsRecorder` in as an extra parameter to `build` (even though it worked in practice!):
```
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 8.403 s <<< FAILURE! - in io.quarkus.it.kafka.streams.KafkaStreamsTest
[ERROR] testKafkaStreams  Time elapsed: 0.007 s  <<< ERROR!
org.junit.jupiter.api.extension.TestInstantiationException: TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [io.quarkus.it.kafka.streams.KafkaStreamsTest]: Failed to start quarkus

...

Caused by: java.lang.NullPointerException
        at java.util.Hashtable.putAll(Hashtable.java:523)
        at io.quarkus.kafka.streams.runtime.KafkaStreamsTopologyManager.getStreamsProperties(KafkaStreamsTopologyManager.java:80)
        at io.quarkus.kafka.streams.runtime.KafkaStreamsTopologyManager.onStart(KafkaStreamsTopologyManager.java:106)
...
```

I can't find anything in the test that seems related, and I can't run it from the IDE as there are a bunch of compilation issues related to `org.apache.maven.wagon.Wagon`.

The implications are that the `processBuildTimeConfig` method can't be changed from being a static init method. From there, the kafkaStreamProperties can't be shared. 

Do you have any suggestions? Do we stick to calling `buildKafkaStreamsProperties` 3 times, or am I missing something? Apologies for taking so much of your time folks - its taking me longer than I anticipated too!

Reply:
I've pushed code that passes the test, but calls `buildKafkaStreamsProperties ` 3 times. Hope its good.